File: Programmers/LV3/67258.cpp
//2020 카카오 인턴십
//보석 쇼핑

#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

vector<int> solution(vector<string> gems) {
    vector<int> answer(2, 0);
    unordered_map<string, int> cur_gems;

    //보석 개수 초기화
    for (string gem : gems) {
        cur_gems[gem];
    }
    int total_gem_cnt = cur_gems.size();
    int start = 0, end = 1;
    int cur_len;

    //맵 초기화
    cur_gems.clear();
    cur_gems[gems[0]]++;

    int min_len = 100001;

    while (start < end) {
        cur_len = end - start;
        if (cur_gems.size() == total_gem_cnt) {
            if (min_len > cur_len) {
                min_len = cur_len;
                answer[0] = start;
                answer[1] = end;
            }

            if (cur_gems[gems[start]] == 1)
                cur_gems.erase(gems[start]);
            else
                cur_gems[gems[start]]--;
            start++;
        }
        else {
            if (end == gems.size())
                break;
            cur_gems[gems[end]]++;
            end++;
        }
    }
    answer[0]++;
    return answer;
}

File: Programmers-py/67258.py
# from collections import defaultdict

# def solution(gems):
#     gem_length = len(gems)
#     total_gem = len(set(gems))
#     count_dict = defaultdict(int)

#     start = 0
#     end = 1
#     count_dict[gems[0]] = 1

#     answer = [-1,-1]
#     gem_range = 100001
#     while start < end:
#         if len(count_dict) == total_gem:
#             cur_range = end - start
#             if cur_range < gem_range:
#                 gem_range = cur_range
#                 answer[0] = start
#                 answer[1] = end
            
#             target_gem = gems[start]
#             if count_dict[target_gem] == 1:
#                 count_dict.pop(target_gem)
#             else:
#                 count_dict[target_gem] -= 1

#             start += 1
#             continue
        
#         if end == gem_length:
#             break
#         count_dict[gems[end]] += 1
#         end += 1

#     answer[0] += 1
#     return answer





from collections import defaultdict

def solution(gems):
    total_gem_cnt = len(set(gems))
    count_dict = defaultdict(int)

    start = 0

    answer = [-1,-1]
    gem_range = 100001
    for end, gem in enumerate(gems, start=1):
        count_dict[gem] += 1
        while len(count_dict) == total_gem_cnt:
            cur_range = end - start
            if cur_range < gem_range:
                gem_range = cur_range
                answer[0] = start
                answer[1] = end

            target_gem = gems[start]
            if count_dict[target_gem] == 1:
                count_dict.pop(target_gem)
            else:
                count_dict[target_gem] -= 1

            start += 1

    answer[0] += 1
    return answer
            


solution(["DIA", "RUBY", "RUBY", "DIA", "DIA", "EMERALD", "SAPPHIRE", "DIA"])
# solution(["ZZZ", "YYY", "NNNN", "YYY", "BBB"])